packages/hashtower.sol/contracts/HashTower.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {PoseidonT3} from "poseidon-solidity/PoseidonT3.sol";
+
+// CAPACITY = W * (W**0 + W**1 + ... + W**(H - 1)) = W * (W**H - 1) / (W - 1)
+uint256 constant H = 24;
+uint256 constant W = 4;
+uint256 constant CAPACITY = 375_299_968_947_540;
+
+// Each HashTower has certain properties and data that will
+// be used to add new items.
+struct HashTowerData {
+    uint256 countOfItems; // count of items being added
+    uint256[H] digests; // digest of each level
+    uint256[H] digestOfDigests; // digest of digests
+}
+
+/// @title HashTower.
+/// @dev The HashTower allows to calculate the digest of digests each time an item is added, ensuring
+/// the integrity of the HashTower.
+library HashTower {
+    uint256 internal constant SNARK_SCALAR_FIELD =
+        21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+    event Add(uint8 indexed level, uint64 indexed lvFullIndex, uint256 value);
+
+    /// @dev Add an item.
+    /// @param self: HashTower data
+    /// @param item: item to be added
+    function add(HashTowerData storage self, uint256 item) public {
```

Comment:
This implementation uses the double nested approach where each level is a hashchain like `H(H(H(item0, item1), item2), item3)`. This implementation is very unfavorable in zk. For example, if i want to prove that `item1` exists in `R = H(H(H(item0, item1), item2), item3)` I need to provide `item0`, `item1`, `item2` and `item3`. This leads to 3x `PoseidonT3` inside the proof.

If I'm proving against a 4x24 structure each proof will require 72 `PoseidonT3`. I think this tradeoff is not worth it when the same proof could be done using 24 `PoseidonT5` if each level is simply `H(item0, item1, item2, item3)`.

The tradeoff is increased gas cost per insertion. In solidity the poseidon function scales slightly non-linearly with input count, `PoseidonT5` is 2.2x more expensive than `PoseidonT3` despite having only 2x inputs. Regardless minimizing constraints is key to building user applications.

I think in the future we could add an option to recursively hash each level, but for now we need to choose one or the other, and I'd rather go with the more zk friendly option. Also bear in mind we'll likely find more optimizations like lazy digest calculation.

Reply:
So reading more the `digestOfDigests` is also a hashchain so the zk scaling should be worse than I wrote above I think? I'd much rather we use the polysum to calculate the digest of the structure itself. Using a constant `R` value is unsafe, but perhaps if we sample an `R**i` value based on the contents of the structure this can be more safe.

Something like take `R**i = levelHash[i-1]` (where `levelHash[-1] = 1`) might work and would keep the favorable gas/constraints property of the polysum.

Even if we don't use a polysum we can form a tree of high arity poseidon hashes. To get `H=24` we can do a `PoseidonT5` hash of `PoseidonT6` hashes. e.g. a shallow tree. This would cost (`PoseidonT5 + PoseidonT6` gas `159k`) (plus the constraints in zk).

Reply:
> If I'm proving against a 4x24 structure each proof will require 72 PoseidonT3. I think this tradeoff is not worth it when the same proof could be done using 24 PoseidonT5 if each level is simply H(item0, item1, item2, item3).

Does that require 4 on-chain PoseidonT5 for the 4 items...?

> The tradeoff is increased gas cost per insertion. In solidity the poseidon function scales slightly non-linearly with input count, PoseidonT5 is 2.2x more expensive than PoseidonT3 despite having only 2x inputs. Regardless minimizing constraints is key to building user applications.

I would like to minimize the gas cost while keeping an acceptable circuit complexity.  However,  I'm not quite familiar with how large the circuit will become unacceptable. Are 10000~20000 constraints acceptable ...?

| W = 4, H = | capacity        | non-linear constraints | zkey size |
|------------|-----------------|------------------------|-----------|
|         12 |        22369620 |                  11738 |   6858502 |
|         16 |      5726623060 |                  15786 |   8860294 |
|         20 |   1466015503700 |                  19866 |  11925126 |
|         24 | 375299968947540 |                  23978 |  13955846 |

> I think in the future we could add an option to recursively hash each level, but for now we need to choose one or the other, and I'd rather go with the more zk friendly option. Also bear in mind we'll likely find more optimizations like lazy digest calculation.

I think both digest() and digestOfDigest() should have much room for improvement. I would like to make them pluggable like the Javascript version, but they are now hardcoded/inlined in the Solidity version for saving gas.
When there are better solutions, we can push a new one.

May we take this 0.0.1 version as the baseline for improvement?

Reply:
> Does that require 4 on-chain PoseidonT5 for the 4 items...?
No, it requires 1 `PoseidonT5`. The 4 items will be an input this.

> I would like to minimize the gas cost while keeping an acceptable circuit complexity. However, I'm not quite familiar with how large the circuit will become unacceptable. Are 10000~20000 constraints acceptable ...?

It's higher than ideal, remember this is just one component in the circuit. So we start with 11k constraints, then add more logic on top. We generally want to stay below 20k constraints total. Also bear in mind some circuits need multiple inclusion proofs.

> May we take this 0.0.1 version as the baseline for improvement?

I think we can merge this into `dev` and not publish to npm to continue working on it.